### PR TITLE
Flesh out C++ FieldValue

### PR DESCRIPTION
**Pull request by wilhuff**
Add a few missing things:

  * `ToString` and `operator<<` to GeoPoint
  * `ToString` and `operator<<` to FieldValue
  * `Hash`
  * Fix `ServerTimestamp`'s `previous_value` to be a `FieldValue`

### CODE REVIEW

File: Firestore/core/include/firebase/firestore/geo_point.h [hunk 1]
```diff
@@ -55,6 +58,16 @@ class GeoPoint {
...
 
+  /**
+   * Returns a string representation of this `GeoPoint` for logging/debugging
+   * purposes.
+   *
+   * Note: the exact string representation is unspecified and subject to change;
+   * don't rely on the format of the string.
+   */
+  std::string ToString() const;
+  friend std::ostream& operator<<(std::ostream& out, const GeoPoint& geo_point);
```
Comment (var-const) [inline Firestore/core/include/firebase/firestore/geo_point.h]: Nit: the implementation doesn't seem to be accessing any private data members, does it have to be a `friend`?

File: Firestore/core/src/firebase/firestore/model/field_value.h [hunk 2]
```diff
@@ -153,6 +146,12 @@ class FieldValue : public util::Comparable<FieldValue> {
     return *blob_value_;
   }
 
+  /**
+   * Returns a string_view of the blob_value(). This can be useful when using
+   * abseil bytewise APIs that accept this type.
+   */
+  absl::string_view blob_value_as_string_view() const;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_value.h]: Very optional: in a similar vein, perhaps `blob_value/blob_size` could be replaced by a single function that returns an `absl::span`?

File: Firestore/core/src/firebase/firestore/model/field_value.h [hunk 3]
```diff
@@ -189,11 +188,16 @@ class FieldValue : public util::Comparable<FieldValue> {
   static FieldValue FromArray(std::vector<FieldValue>&& value);
   static FieldValue FromMap(const Map& value);
   static FieldValue FromMap(Map&& value);
+  static FieldValue FromMap(
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_value.h]: Hmm, `SortedMap`'s constructor taking an `initializer_list` is not `explicit`, so it seems like `FromMap({"foo", {}})` should have worked already. Was it not the case?

File: Firestore/core/src/firebase/firestore/model/field_value.h [hunk 4]
```diff
@@ -277,13 +281,41 @@ class ObjectValue : public util::Comparable<ObjectValue> {
...
+  friend std::ostream& operator<<(std::ostream& os,
+                                  const ServerTimestamp& value);
+
+  size_t Hash() const;
+};
+
+struct ReferenceValue {
+  DocumentKey reference;
+  // Does not own the DatabaseId instance.
+  const DatabaseId* database_id;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_value.h]: Very optional: add `= nullptr` initialization?

File: Firestore/core/src/firebase/firestore/model/field_value.h [hunk 5]
```diff
@@ -189,11 +188,16 @@ class FieldValue : public util::Comparable<FieldValue> {
...
   static FieldValue FromMap(Map&& value);
+  static FieldValue FromMap(
+      std::initializer_list<std::pair<std::string, FieldValue>> value);
 
   size_t Hash() const;
 
   util::ComparisonResult CompareTo(const FieldValue& rhs) const;
 
+  std::string ToString() const;
+  friend std::ostream& operator<<(std::ostream& os, const FieldValue& value);
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_value.h]: Same question about `friend`ship here.

File: Firestore/core/src/firebase/firestore/model/field_value.cc [hunk 6]
```diff
@@ -438,6 +500,45 @@ ComparisonResult FieldValue::CompareTo(const FieldValue& rhs) const {
...
+    case Type::GeoPoint:
+      return util::ToString(*geo_point_value_);
+    case Type::Array:
+      return util::ToString(*array_value_);
+    case Type::Object:
+      return util::ToString(*object_value_);
+    default:
+      HARD_FAIL("Unsupported type %s", type());
+      // return false if assertion does not abort the program. We will say
+      // each unsupported type takes only one value thus everything is equal.
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_value.cc]: `We will say each unsupported type takes only one value thus everything is equal.`

I'm not sure I understand this comment, can you rephrase a bit?

File: Firestore/core/src/firebase/firestore/model/field_value.cc [hunk 7]
```diff
@@ -438,6 +500,45 @@ ComparisonResult FieldValue::CompareTo(const FieldValue& rhs) const {
...
+    case Type::Reference:
+      return util::ToString(*reference_value_);
+    case Type::GeoPoint:
+      return util::ToString(*geo_point_value_);
+    case Type::Array:
+      return util::ToString(*array_value_);
+    case Type::Object:
+      return util::ToString(*object_value_);
+    default:
+      HARD_FAIL("Unsupported type %s", type());
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_value.cc]: Question: `HARD_FAIL` is used instead of `UNREACHABLE` because of the more informative error message, right?

File: Firestore/core/test/firebase/firestore/model/field_value_test.cc [hunk 8]
```diff
@@ -36,6 +38,53 @@ const uint8_t* Bytes(const char* value) {
...
+  EXPECT_EQ("foo", FieldValue::FromString("foo").ToString());
+
+  // Bytes escaped as hex
+  const char* hi = "HI";
+  auto blob = FieldValue::FromBlob(reinterpret_cast<const uint8_t*>(hi), 2);
+  EXPECT_EQ("<4849>", blob.ToString());
+
+  DatabaseId database_id("p", "d");
+  auto ref = FieldValue::FromReference(Key("foo/bar"), &database_id);
+  EXPECT_EQ("Reference(key=foo/bar)", ref.ToString());
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/field_value_test.cc]: Very optional: perhaps add single quotes to the string representation?
```
"Reference(key='foo/bar')"
```

File: Firestore/core/src/firebase/firestore/model/field_value.h (hunk 5 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/field_value.h]: Same response. In this case moving it out of the class feels messy since there are so many other classes in this file.

File: Firestore/core/src/firebase/firestore/model/field_value.h (hunk 4 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/field_value.h]: Done.

File: Firestore/core/include/firebase/firestore/geo_point.h (hunk 1 above)
Comment (var-const) [reply Firestore/core/include/firebase/firestore/geo_point.h]: This rationale makes sense to me. Feel free to keep as it was, and sorry about the trouble.

File: Firestore/core/test/firebase/firestore/model/field_value_test.cc (hunk 8 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/model/field_value_test.cc]: The other ports don't do this, so I think I'm going to pass on this.
